### PR TITLE
gh-105288: wake up exit waiters after sub-process exits

### PR DESCRIPTION
**Pull request by keuin**
* Issue: gh-105288

### CODE REVIEW

**Comment by gvanrossum** [comment]
@kumaraditya303 Once you are convinced of the validity of the issue (which I expect you will be), will you review ad merge this PR? I was on vacation for 10 days and am way behind.

**Comment by kumaraditya303** [comment]
As evident by the failing CI this is wrong.

**Comment by keuin** [comment]
I will fix my test in these days. Sorry because I'm not familiar with the test framework and testing a frozen event loop is much harder than usual.

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]
See my explanation on the issue

**Comment by Hels15** [comment]
What's going on here?

File: Lib/test/test_asyncio/test_subprocess.py [hunk 1]
```diff
@@ -847,6 +849,35 @@ async def main() -> None:
...
+            t = asyncio.create_task(whoami())
+            await asyncio.wait([0, t])
+
+        def _main():
+            nonlocal finished
+            try:
+                # We have to run the event loop in a separate thread,
+                # to make sure we can fail in the main thread
+                # and successfully notify the test driver.
+                self.loop.run_until_complete(main())
```
Comment (graingert) [inline Lib/test/test_asyncio/test_subprocess.py:871]: the problem is in asyncio.run - so this doesn't actually test the issue

**Comment by keuin** [comment]
I think this issue is a duplicate of others. The misuse of argument shows some deep resource leak, but is hard to find the root cause.
I'm glad to see others want to fix the similar issue. I want to wait to see if it can fix this too.

**Pull request closed without merging**